### ISSUE TITLE
Add a powershell engine

### ISSUE THREAD

**Issue by cderv** [yihui/knitr#1932]
Why : 
* bash or sh is not always accessible on windows 
* It would be good to have a Windows equivalent

This could be made easier after working on a generic engine https://github.com/yihui/knitr/pull/1829

Maybe usefull command : `xfun:::powershell()`

Use case : 
* Would allow to use reprex to build reproducible example in powershell
* Would allow to write tutorials for Powershell command

**Comment by znmeb** [comment]
Is there documentation on how to add a language engine for arbitrary compiled or interpreted languages? I'm interested in learning Rust and there doesn't seem to be a language engine for it.

**Comment by cderv** [comment]
There is already a PR for a Rust engine. See https://github.com/yihui/knitr/pull/1823

However, we are waiting before adding more engine as we plan on adding a feature for a Generic engine (based on proposal #1829). 

You can try the Rust PR though in the meantime!

About registering a custom engine, you can see about it in the R Markdown Cookbook
https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html

**Comment by yihui** [comment]
@cderv Does ```` ```{exec, command="powershell"}```` work now? If it works, we can add register the `powershell` engine (or `ps` for short).

> Is there documentation on how to add a language engine for arbitrary compiled or interpreted languages?

@znmeb Please see https://github.com/yihui/knitr/pull/1829#issuecomment-1011801414

**Comment by cderv** [comment]
> If it works, we can add register the powershell engine (or ps for short).

I believe we'll need to register an engine to bring a few tweaks. 

powershell expect file with extension `.ps1`. Example with `pwsh` (newer powershell) gives this error
````
Processing -File 'pwsh24c04384463d.pwsh' failed because the file does not have a '.ps1' extension. Specify a valid PowerShell script file name, and then try again.
````
  
`powershell` will also error but differently. 


This works 
````markdown

```{exec, command = "pwsh"}
#| engine.opts:
#|    ext: ps1
$env:PATH -split ";"
```

````

However, with `powershell` it seems it require explicit relative path

* Works: `powershell.exe .\powershell24c01c3480a.ps1`
* Does not work: 

This works though
````markdown
```{exec, command = "powershell"}
#| engine.opts:
#|    args1: -File
#|    ext: ps1
$env:PATH -split ";"
```
````

So we could add a `ps` engine with those tweaks probably. I can make a PR

**Comment by Shaunson26** [comment]
I am curious about this. I've had to do learn some basic powershell stuff for work. I thought of keeping my notes with Rmarkdown document. I just copied code from 15.1 of the Rmarkdown cookbook. I'm curious about

1.  inline code vs -file argument (from above posts)
2. keeping the environment across chunks
<pre>
```{r}
# Allow stderr chunk option
knitr::knit_engines$set(ps = function(options) {
  code <- paste(options$code, collapse = '\n')
  out  <- system2(
    'powershell', c('-c', shQuote(code)), 
    stdout = TRUE, stderr = if (is.null(options$stderr)) TRUE
  )
  knitr::engine_output(options, code, out)
})
```
</pre>
<pre>
```{ps}
write-output "Hello"
$this = "that"
write-output "this equals $this"
$this
```
</pre>

<pre>
## Hello
## this equals that
## that
</pre>

<pre>
```{ps}
# each chunk runs in it's own environment, $this from above is not available (expected)
$this
```
</pre>

<pre>
</pre>

<pre>
```{ps}
# R warning + PS stderr
write-error "An error"
```
</pre>

<pre>
Warning in system2("powershell", c("-c", shQuote(code)), stdout = TRUE,  :
  running command '"powershell" -c "write-error \"An error\""' had status 1
## write-error "An error" : An error
##     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
##     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
##
</pre>

<pre>
```{ps, warning = FALSE, stderr = FALSE}
# Hide both
write-error "An error"
```
</pre>

<pre>
</pre>

**Comment by Shaunson26** [comment]
when you wrote `powershell.exe .\powershell24c01c3480a.ps1` above, is that file something created from knitr i.e. the code chunk being saved to file and then executed? or something the user is providing? I'm likely missing some intricacies of knitr here..

**Comment by cderv** [comment]
> is that file something created from knitr i.e. the code chunk being saved to file and then executed? 

Yes - **knitr** will put the content of a chunk to a file and execute it. That is how the `exec` engine is working.

**Comment by yihui** [comment]
Thanks for the suggestion, @cderv! Added in #2504 (will be in the next release).

**Usage:** mark a chunk with the `ps` engine and write PowerShell as usual:

````markdown
```{ps}
Write-Output "Hello from PowerShell"
$PSVersionTable.PSVersion
```
````

**Details:**
- The chunk is written to a temporary `.ps1` script and run with `powershell -ExecutionPolicy Bypass -File`. The `.ps1` extension and `-File` are required by PowerShell (as you found); `-ExecutionPolicy Bypass` keeps the unsigned temp script from being blocked by Windows' default policy.
- It's a thin wrapper over the generic `exec` engine, so you can point it at the newer cross-platform PowerShell with `engine.opts = list(command = "pwsh")`, or at an explicit path via `engine.path`.
- Code is highlighted as `powershell`.

As you noted upthread, each chunk runs via `system2()` in its own process, so state isn't shared across chunks — that's expected.

**Comment by cderv** [comment]
Thank you:!
